fsharp-generator/src/helpers.ts

```diff
@@ -0,0 +1,575 @@
+        realBaseType = realBaseType.type;
+      }
+
+      return `
+              if (${valueVar}.IsNone) then
+                resultWriter_.WriteNullValue()
+              else
+                ${encodeType(realBaseType, `${valueVar}.Value`, path, suffix, isRef)}
+              `
+        .replace(/\n {14}/gu, "\n")
+        .trim();
+    }
+
+    case TypeReference:
+      return encodeType((type as TypeReference).type, valueVar, path, suffix, isRef);
+    case EnumType:
+      return `Encode${type.name} ${valueVar} resultWriter_ ${path}`;
+    case StructType:
+      return `Encode${type.name} ${valueVar} resultWriter_ ${path}`;
+    case JsonPrimitiveType:
+      return `${valueVar}.WriteTo(resultWriter_)`;
+    case ArrayType: {
+      return `
+              resultWriter_.WriteStartArray()
+              for i${suffix} in 0..${valueVar}.Length - 1 do
+                ${encodeType((type as ArrayType).base, `${valueVar}.[i${suffix}]`, `${path}`, suffix + 1).replace(/\n/gu, "\n                ")}
+              resultWriter_.WriteEndArray()
+              `
+        .replace(/\n {14}/gu, "\n")
+        .trim();
+    }
+
+    default:
+      throw new Error(`BUG: encodeType with ${type.constructor.name}`);
+  }
+}
+
+export function generateStruct(struct: StructType): string {
+  return `
+type ${struct.name} = {
+  ${struct.fields.map(field => `${capitalize(field.name)}: ${generateTypeName(field.type)}`).join(";\n  ")}
+} with
+  static member Create (${struct.fields.map(field => `${ident(field.name)}: ${generateTypeName(field.type)}`).join(", ")}): ${struct.name} =
+    { ${struct.fields.map(field => `${capitalize(field.name)} = ${ident(field.name)}`).join("; ")} }
+
+let Decode${struct.name} (json_: JsonElement) (path_: string): ${struct.name} =
+  if (json_.ValueKind <> JsonValueKind.Object) then raise (FatalException($"'{path_}' must be an object."))
+  ${struct.fields
+    .map(
+      field =>
+        `
+  let ${field.name}Json_ = decodeJsonElementStrict ${JSON.stringify(field.name)} json_ $"{path_}.${field.name}"
+  let ${ident(field.name)} =
+    ${decodeType(field.type, `${field.name}Json_`, `$"{path_}.${field.name}"`, ident(field.name)).replace(/\n/gu, "\n  ")}
+  `,
+    )
+    .join("")}
+  { ${struct.fields.map(field => `${capitalize(field.name)} = ${ident(field.name)}`).join("; ")} }
+
+let Encode${struct.name} (obj_: ${struct.name}) (resultWriter_: Utf8JsonWriter) (path_: string) =
+  resultWriter_.WriteStartObject()
+  ${struct.fields
+    .map(
+      field =>
+        `
+  resultWriter_.WritePropertyName(${JSON.stringify(field.name)})
+  ${encodeType(field.type, `obj_.${capitalize(field.name)}`, `$"{path_}.${field.name}"`).replace(/\n/gu, "\n  ")}`,
+    )
+    .join("\n")}
+  resultWriter_.WriteEndObject()`;
+}
+
+export function generateEnum(type: EnumType): string {
+  return `
+type ${type.name} =
+  ${type.values.map(({ value }) => `| ${capitalize(value)}`).join("\n  ")}
```

Comment:
`value` agora tem duas propriedades, uma chamada `value` e outra chamada `struct`. Tem que usar `value.value` ali. E se struct for != null, sugiro lançar um erro de "enums with embedded structs aren't supported for F# yet".